### PR TITLE
enhancement restart policy

### PR DESCRIPTION
**Pull request by junnplus**
Signed-off-by: Ye Sijun <junnplus@gmail.com>

Fixes: https://github.com/containerd/nerdctl/issues/945

### CODE REVIEW

File: cmd/nerdctl/run_restart.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+			return nil, err
+		}
+		ver, err := semver.NewVersion(version.Version)
+		if err != nil {
+			return nil, err
+		}
+		ver163 := semver.MustParse("v1.6.3")
+		if ver.LessThan(ver163) {
+			return nil, fmt.Errorf("containerd version %s does not support restart policy %s", version.Version, policySlice[0])
+		}
```
Comment (Junnplus) [inline cmd/nerdctl/run_restart.go:47]: For the main branch, it seems to always be v1.6.0-xxx.
```
$ containerd --version
containerd github.com/containerd/containerd v1.6.0-290-g0c2b6a102 0c2b6a102a3a40819c28234ac41eec6c919d6b2f
```

Comment (AkihiroSuda) [reply cmd/nerdctl/run_restart.go:47]: Please check the plugin metadata, not the daemon version

https://github.com/containerd/containerd/blob/7f8506a4f96d5de14a94d47b28265ac1652a61c1/runtime/restart/monitor/monitor.go#L76

Comment (Junnplus) [reply cmd/nerdctl/run_restart.go:47]: Makes sense

File: cmd/nerdctl/run_restart_linux_test.go [hunk 2]
```diff
@@ -87,3 +89,48 @@ func TestRunRestart(t *testing.T) {
...
+	poll.WaitOn(t, check, poll.WithDelay(100*time.Microsecond), poll.WithTimeout(60*time.Second))
+	inspect := base.InspectContainer(tID)
+	assert.Equal(t, inspect.RestartCount, 2)
+}
+
+func TestRunRestartWithUnlessStopped(t *testing.T) {
+	base := testutil.NewBase(t)
+	if base.Target == testutil.Nerdctl && os.Getenv("CONTAINERD_VERSION") != "main" {
+		t.Skip("only test containerd on main version")
+	}
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_restart_linux_test.go:118]: Could you use `testutil.RequireContainerdPlugin()` ? (being added in https://github.com/containerd/nerdctl/pull/726)

File: cmd/nerdctl/run_restart.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	if len(res.Plugins) == 0 {
+		return nil, fmt.Errorf("no restart plugin found")
+	}
+	restartPlugin := res.Plugins[0]
+	capabilities := restartPlugin.Capabilities
+	if len(capabilities) == 0 {
+		capabilities = []string{"always"}
+	}
+	if !strutil.InStringSlice(capabilities, policySlice[0]) {
+		return nil, fmt.Errorf("unsupported restart policy %q, supported policy are: %q", policySlice[0], restartPlugin.Capabilities)
```
Comment (fahedouch) [inline cmd/nerdctl/run_restart.go:49]: typo: supported `policies` are

File: cmd/nerdctl/run_restart.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/runtime/restart"
+	"github.com/containerd/nerdctl/pkg/strutil"
+)
+
+func generateRestartOpts(ctx context.Context, client *containerd.Client, restartFlag, logURI string) ([]containerd.NewContainerOpts, error) {
+	policySlice := strings.Split(restartFlag, ":")
+	switch policySlice[0] {
+	case "", "no":
```
Comment (fahedouch) [inline cmd/nerdctl/run_restart.go:33]: empty policy should be supported too ?
  https://github.com/containerd/containerd/blob/main/runtime/restart/restart.go#L85

Comment (junnplus) [reply cmd/nerdctl/run_restart.go:33]: For containerd, empty policy is to be compatible with scenes without policy label.

File: go.mod [hunk 5]
```diff
@@ -11,11 +11,11 @@ require (
 	github.com/containerd/containerd v1.6.3
 	github.com/containerd/continuity v0.3.0
 	github.com/containerd/go-cni v1.1.5
-	github.com/containerd/imgcrypt v1.1.4
-	github.com/containerd/stargz-snapshotter v0.11.4
+	github.com/containerd/imgcrypt v1.1.5-0.20220421044638-8ba028dca028
+	github.com/containerd/stargz-snapshotter v0.11.1
```
Comment (AkihiroSuda) [inline go.mod:15]: Why rolled back?

**Comment by AkihiroSuda** [comment]
Thanks, the code looks good but go.mod seems to have unneeded changes accidentally (?)

File: go.mod [hunk 6]
```diff
@@ -11,11 +11,11 @@ require (
 	github.com/containerd/containerd v1.6.3
 	github.com/containerd/continuity v0.3.0
 	github.com/containerd/go-cni v1.1.5
-	github.com/containerd/imgcrypt v1.1.4
+	github.com/containerd/imgcrypt v1.1.5-0.20220421044638-8ba028dca028
```
Comment (AkihiroSuda) [inline go.mod:14]: Why do we need updating imgcrypt and typeurl?

File: go.mod [hunk 7]
```diff
@@ -160,11 +166,13 @@ require (
...
+	google.golang.org/genproto v0.0.0-20220426171045-31bebdecfb46 // indirect
+	google.golang.org/grpc v1.46.0 // indirect
+	google.golang.org/protobuf v1.28.0 // indirect
 	gopkg.in/square/go-jose.v2 v2.5.1 // indirect
 )
 
 // Temporary fork for avoiding importing patent-protected code: https://github.com/hashicorp/golang-lru/issues/73
 replace github.com/hashicorp/golang-lru => github.com/ktock/golang-lru v0.5.5-0.20211029085301-ec551be6f75c
+
+replace github.com/containerd/containerd => github.com/containerd/containerd v1.6.1-0.20220428184543-bb8b134a1797
```
Comment (AkihiroSuda) [inline go.mod:178]: Why do we need replace() ? Can we just specify this revision in require() ?

Comment (junnplus) [reply go.mod:178]: I tried it but go get fails :(
```
 go get github.com/containerd/containerd@main
go: github.com/containerd/containerd@main (v1.6.1-0.20220504173539-885468815ef7) requires github.com/containerd/containerd@v1.6.1, not github.com/containerd/containerd@main (v1.6.1-0.20220504173539-885468815ef7)
```

File: go.mod (hunk 6 above)
Comment (junnplus) [reply go.mod:14]: I'm not sure, I just replaced containerd.

**Review by AkihiroSuda — APPROVED** [review]

**Pull request merged**